sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala

```diff
@@ -161,6 +161,267 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
+          ("\"listagg(DISTINCT a, NULL) " +
+          "OVER (ORDER BY a ASC NULLS FIRST RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW)\"")),
+        context = ExpectedContext(
+          fragment = "listagg(distinct a) over (order by a)",
+          start = 7,
+          stop = 43))
+
+      checkAnswer(
+        sql("select listagg(distinct a) within group (order by a DESC) from df"),
+        Row("ba") :: Nil)
+
+      checkError(
+        exception = intercept[AnalysisException] {
+          sql("select listagg(distinct a) within group (order by b) from df")
+        },
+        condition = "FUNCTION_AND_ORDER_EXPRESSION_MISMATCH",
+        parameters = Map(
+          "functionName" -> "`listagg`",
+          "functionArgs" -> "\"a\"",
+          "orderExpr" -> "\"b\""),
+        context = ExpectedContext(
+          fragment = "listagg(distinct a) within group (order by b)",
+          start = 7,
+          stop = 51))
+
+      checkError(
+        exception = intercept[AnalysisException] {
+          sql("select listagg(distinct a) within group (order by a, b) from df")
+        },
+        condition = "FUNCTION_AND_ORDER_EXPRESSION_MISMATCH",
+        parameters = Map(
+          "functionName" -> "`listagg`",
+          "functionArgs" -> "\"a\"",
+          "orderExpr" -> "\"a\", \"b\""),
+        context = ExpectedContext(
+          fragment = "listagg(distinct a) within group (order by a, b)",
+          start = 7,
+          stop = 54))
+
+      Seq((1, true), (2, false), (3, false)).toDF("a", "b").createOrReplaceTempView("df2")
+
+      checkAnswer(
+        sql("select listagg(a), listagg(b, ',') from df2"),
+        Row("123", "true,false,false") :: Nil)
+    }
+  }
+
+  test("listagg collation test") {
+    checkAnswer(
+      sql("select listagg(c1) within group (order by c1 collate utf8_binary)" +
+        " from values ('a'), ('A'), ('b'), ('B') as t(c1)"),
+      Row("ABab") :: Nil)
+
+    checkAnswer(
+      sql("select listagg(c1) within group (order by c1 collate utf8_lcase)" +
+        " from values ('a'), ('A'), ('b'), ('B') as t(c1)"),
+      Row("aAbB") :: Nil)
+
+    checkAnswer(
+      sql("select listagg(DISTINCT c1 collate utf8_binary)" +
+        " from values ('a'), ('A'), ('b'), ('B') as t(c1)"),
+      Row("aAbB") :: Nil)
+
+    checkAnswer(
+      sql("select listagg(DISTINCT c1 collate utf8_lcase)" +
+        " from values ('a'), ('A'), ('b'), ('B') as t(c1)"),
+      Row("ab") :: Nil)
+
+    checkAnswer(
+      sql("select listagg(DISTINCT c1 collate utf8_lcase)" +
+        " within group (order by c1 collate utf8_lcase)" +
+        " from values ('a'), ('B'), ('b'), ('A') as t(c1)"),
+      Row("aB") :: Nil)
+
+    checkError(
+      exception = intercept[AnalysisException] {
+        sql(
+          """select listagg(DISTINCT c1 collate utf8_lcase)
+            | within group (order by c1 collate utf8_binary)
+            | from values ('a'), ('b'), ('A'), ('B') as t(c1)""".stripMargin)
```

Comment:
do we still need these tests if they are already in golden files?

Reply:
test duplication is a bad thing, let's remove them. We should make the golden file test easier to debug if you find problems. We won't and shouldn't duplicate all the golden file tests in scala.